File: packages/shared/src/types/cds.ts
/**
 * Clinical Decision Support (CDS) types.
 * Used by the CDS evaluation engine to return alerts for prescriptions,
 * vitals, and lab orders.
 */

/** Severity level for a CDS alert. */
export type AlertSeverity = "critical" | "high" | "moderate" | "low" | "info";

/** Category describing the type of clinical decision support rule. */
export type AlertCategory =
  | "drug_interaction"
  | "drug_allergy"
  | "drug_condition"
  | "drug_age"
  | "drug_pregnancy"
  | "drug_duplicate"
  | "drug_dose"
  | "vital_critical"
  | "vital_trend"
  | "lab_critical"
  | "lab_trend"
  | "lab_organ_dysfunction"
  | "sepsis_risk"
  | "diagnosis_guideline"
  | "duplicate_order"
  | "readmission_risk"
  | "clinical_pathway"
  | "formulary_check";

/** Action the CDS system recommends for the alert. */
export type AlertAction = "block" | "warn" | "suggest" | "inform";

/** A single CDS alert returned by the evaluation engine. */
export interface CDSAlert {
  alert_id: string;
  severity: AlertSeverity;
  category: AlertCategory;
  action: AlertAction;
  title: string;
  message: string;
  source_rule: string;
  affected_items: string[];
  evidence: Record<string, unknown>;
  overridable: boolean;
  requires_reason: boolean;
}

/** Response from the CDS evaluation endpoint. */
export interface CDSEvaluationResponse {
  alerts: CDSAlert[];
  is_blocked: boolean;
  blocking_alerts: CDSAlert[];
  evaluation_time_ms: number;
}

/** Request payload for evaluating prescription CDS rules. */
export interface CDSPrescriptionRequest {
  patient_id: string;
  drug_name: string;
  encounter_id?: string | null;
}

/** Request payload for evaluating vitals CDS rules. */
export interface CDSVitalsRequest {
  patient_id: string;
  vitals: Record<string, number | string | null>;
}

/** Request payload for evaluating lab result CDS rules. */
export interface CDSLabRequest {
  patient_id: string;
  test_code: string;
  result_numeric: number | null;
}

File: services/api-gateway/app/schemas/cds.py
"""
Pydantic schemas for Clinical Decision Support (CDS) API endpoints.
"""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class CDSPrescriptionRequest(BaseModel):
    """
    Request to evaluate CDS rules for a prescription.

    @param patient_id: Patient UUID
    @param drug_name: Drug being prescribed
    @param encounter_id: Optional current encounter UUID
    """

    patient_id: uuid.UUID
    drug_name: str = Field(..., min_length=1, max_length=255)
    encounter_id: uuid.UUID | None = None

    model_config = ConfigDict(from_attributes=True)


class CDSVitalsRequest(BaseModel):
    """
    Request to evaluate CDS rules for vital signs.

    @param patient_id: Patient UUID
    @param vitals: Vital sign values (e.g. heart_rate, systolic_bp, etc.)
    """

    patient_id: uuid.UUID
    vitals: dict[str, float | int | str | None] = Field(
        ..., description="Vital sign key-value pairs"
    )

    model_config = ConfigDict(from_attributes=True)


class CDSLabRequest(BaseModel):
    """
    Request to evaluate CDS rules for a lab result.

    @param patient_id: Patient UUID
    @param test_code: Lab test code (e.g. WBC, HGB, CREAT)
    @param result_numeric: Numeric result value, or None for qualitative results
    """

    patient_id: uuid.UUID
    test_code: str = Field(..., min_length=1, max_length=50)
    result_numeric: float | None = None

    model_config = ConfigDict(from_attributes=True)


class CDSAlertResponse(BaseModel):
    """
    A single CDS alert for API output.

    @param alert_id: Unique alert identifier
    @param severity: Alert severity level (critical, high, moderate, low, info)
    @param category: Alert classification
    @param action: Recommended clinician action (block, warn, suggest, inform)
    @param title: Short alert title
    @param message: Detailed alert description
    @param source_rule: Rule that triggered this alert
    @param affected_items: List of affected drug names, test codes, etc.
    @param evidence: Supporting data (thresholds, references)
    @param overridable: Whether the clinician can override this alert
    @param requires_reason: Whether override requires a documented reason
    @param created_at: When the alert was generated
    """

    alert_id: str
    severity: str
    category: str
    action: str
    title: str
    message: str
    source_rule: str
    affected_items: list[str] = Field(default_factory=list)
    evidence: dict[str, Any] = Field(default_factory=dict)
    overridable: bool = True
    requires_reason: bool = False
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CDSEvaluationApiResponse(BaseModel):
    """
    Response from CDS evaluation API endpoint.

    @param alerts: List of triggered alerts
    @param is_blocked: Whether the action should be blocked
    @param blocking_alerts: Subset of alerts that cause blocking
    @param evaluation_time_ms: Time taken for evaluation in milliseconds
    """

    alerts: list[CDSAlertResponse]
    is_blocked: bool = False
    blocking_alerts: list[CDSAlertResponse] = Field(default_factory=list)
    evaluation_time_ms: float = 0.0

    model_config = ConfigDict(from_attributes=True)
